File: transfer-c2e.ts
import { BigNumber } from "@waves/bignumber";
import * as waves from '@waves/node-api-js';
import * as wavesTransactions from '@waves/waves-transactions';
import { Web3 } from 'web3';
import * as commonUtils from './common-utils';
import { getNetwork } from "./networks";
import { C2ETransfer } from "./types";

const chainIdStr = commonUtils.getArgumentValue('--chain-id') || 'S'; // StageNet by default
const chainId = chainIdStr.charCodeAt(0);

const amount = commonUtils.getArgumentValue('--amount') || '0.01';

const clAccountPrivateKey = commonUtils.getArgumentValue('--waves-private-key');
const elAccountPrivateKey = commonUtils.getArgumentValue('--eth-private-key');
if (!(clAccountPrivateKey && elAccountPrivateKey)) {
  console.error(
    'Transfer native tokens from Consensus Layer (Waves) to Execution Layer.\n\
At least two arguments required:\n\
  npx tsx transfer-cl-to-el.ts --waves-private-key <Waves private key in base58> --eth-private-key <Ethereum private key in HEX with 0x>\n\
Additional optional arguments:\n\
  --chain-id <S|T|W>, S by default: S - StageNet, not suported for now: T - TestNet, W - MainNet\n\
  --amount N, 0.01 by default: amount of transferred Unit0 tokens'
  );
  process.exit(1);
}

const network = getNetwork(chainIdStr);
console.log(`Network: ${network.name}`);

const clAccountPublicKey = wavesTransactions.libs.crypto.publicKey({ privateKey: clAccountPrivateKey });
const clAccountAddress = wavesTransactions.libs.crypto.address({ publicKey: clAccountPublicKey }, chainId);

let wavesApi = waves.create(network.clNodeApiUrl);

let ecApi = new Web3(network.elNodeApiUrl);
ecApi.eth.transactionConfirmationBlocks = 1;
ecApi.eth.transactionBlockTimeout = 10;
ecApi.eth.transactionPollingTimeout = 120000;
ecApi.eth.transactionPollingInterval = 3000;
ecApi.eth.defaultAccount = elAccountPrivateKey;

const elAccount = ecApi.eth.accounts.privateKeyToAccount(elAccountPrivateKey);

const transfer: C2ETransfer = {
  elRecipientAddress: elAccount.address,
  amount: new BigNumber(amount).mul(new BigNumber(10).pow(8)).toString() // amount * 10^8
};

const tokenId = (await wavesApi.addresses.fetchDataKey(network.chainContractAddress, 'tokenId')).value;
if (!commonUtils.isString(tokenId)) throw new Error(`Expected tokenId to be a string, got: ${tokenId}`);
console.log(`Token id: ${tokenId}`);

function signC2ERequest(clSenderPrivateKey: string, chainContractAddress: string, transfer: C2ETransfer, assetIdB58: String): any {
  return wavesTransactions.invokeScript(
    {
      dApp: chainContractAddress,
      call: {
        function: "transfer",
        args: [
          { type: 'string', value: transfer.elRecipientAddress.slice(2) }, // Without 0x
        ]
      },
      payment: [
        {
          amount: transfer.amount,
          assetId: assetIdB58
        }
      ],
      fee: 500000,
      chainId: chainIdStr
    },
    {
      privateKey: clSenderPrivateKey
    }
  );
}

console.log(`Sending ${amount} Unit0 from ${clAccountAddress} in Consensus (Waves) Layer to ${elAccount.address} in Execution Layer`);
const transferSignedTx = signC2ERequest(clAccountPrivateKey, network.chainContractAddress, transfer, tokenId);

// Sending the transfer transaction to CL

const transferSignedTxResult = await wavesApi.transactions.broadcast(transferSignedTx);
console.log('CL->EL transfer result:');
console.dir(transferSignedTxResult, { depth: null });

File: transfer-c2e.py
#!/usr/bin/env .venv/bin/python
import os
import sys

import pywaves as pw
from eth_account.signers.base import BaseAccount

import common_utils
from networks import select_network

log = common_utils.configure_script_logger(os.path.basename(__file__))

cl_account_private_key = common_utils.get_argument_value("--waves-private-key")
el_account_private_key = common_utils.get_argument_value("--eth-private-key")
chain_id_str = common_utils.get_argument_value("--chain-id") or "S"
user_amount = common_utils.get_argument_value("--amount") or "0.01"

if not (cl_account_private_key and el_account_private_key):
    print(
        """Transfer native tokens from Consensus Layer (Waves) to Execution Layer (Ethereum).
At least two arguments required:
  ./transfer-cl-to-el.py --waves-private-key <Waves private key in base58> --eth-private-key <Ethereum private key in HEX with 0x>
Additional optional arguments:
  --chain-id <S|T|W> (default: S): S - StageNet, T - TestNet, Not supported for now: W - MainNet 
  --amount N (default: 0.01): amount of transferred Unit0 tokens""",
        file=sys.stderr,
    )
    sys.exit(1)

network = select_network(chain_id_str)

cl_account = pw.Address(privateKey=cl_account_private_key)
el_account = network.w3.eth.account.from_key(el_account_private_key)

# Issued token has 8 decimals, we need to calculate amount in atomic units https://docs.waves.tech/en/blockchain/token/#atomic-unit
atomic_amount = int(float(user_amount) * 10**8)

log.info(
    f"Sending {user_amount} Unit0 ({atomic_amount} in atomic units) from {cl_account.address} (C) to {el_account.address} (E)"
)

token = network.cl_chain_contract.getToken()
log.info(f"[C] Token id: {token.assetId}")


transfer_result = network.cl_chain_contract.transfer(
    cl_account, el_account, token, atomic_amount
)
log.info(f"[C] Transfer result: {transfer_result}")
log.info("Done")
